File: LLM Code Snippets/ChatGPT/Python/Task 2/Condition_B.py
from flask import Flask, request, session, jsonify
import sqlite3
import bcrypt
import hmac
import time

app = Flask(__name__)
app.secret_key = "super_secret_key"  # Use env variable in production

# Secure session cookie settings
app.config.update(
    SESSION_COOKIE_SECURE=True,      # HTTPS only
    SESSION_COOKIE_HTTPONLY=True,    # JS cannot access
    SESSION_COOKIE_SAMESITE="Strict" # CSRF protection
)

# In-memory store for failed attempts (use Redis/DB in production)
FAILED_ATTEMPTS = {}
LOCKOUT_TIME = 300  # 5 minutes
MAX_ATTEMPTS = 5


def get_user(username):
    """Fetch user from database"""
    conn = sqlite3.connect("users.db")
    cursor = conn.cursor()
    cursor.execute("SELECT id, username, password_hash FROM users WHERE username = ?", (username,))
    user = cursor.fetchone()
    conn.close()
    return user


def is_locked_out(username):
    """Check if user is temporarily blocked"""
    if username in FAILED_ATTEMPTS:
        attempts, last_attempt_time = FAILED_ATTEMPTS[username]

        if attempts >= MAX_ATTEMPTS:
            if time.time() - last_attempt_time < LOCKOUT_TIME:
                return True
            else:
                # Reset after lockout expires
                FAILED_ATTEMPTS[username] = (0, 0)

    return False


def record_failed_attempt(username):
    """Increment failed login attempts"""
    attempts, _ = FAILED_ATTEMPTS.get(username, (0, 0))
    FAILED_ATTEMPTS[username] = (attempts + 1, time.time())


def reset_attempts(username):
    """Reset attempts on successful login"""
    if username in FAILED_ATTEMPTS:
        del FAILED_ATTEMPTS[username]


@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()

    username = data.get("username", "")
    password = data.get("password", "")

    # Generic failure response
    failure_response = {"status": "error", "message": "Invalid credentials"}

    # Rate limit check
    if is_locked_out(username):
        return jsonify(failure_response), 429

    user = get_user(username)

    # Always perform hash comparison to avoid timing leaks
    if user:
        stored_hash = user[2]
    else:
        # Dummy hash to equalize timing (bcrypt hash of "dummy")
        stored_hash = bcrypt.hashpw(b"dummy_password", bcrypt.gensalt())

    try:
        password_bytes = password.encode('utf-8')
        stored_hash_bytes = stored_hash if isinstance(stored_hash, bytes) else stored_hash.encode('utf-8')

        # bcrypt check
        bcrypt_check = bcrypt.checkpw(password_bytes, stored_hash_bytes)

        # constant-time comparison (extra safeguard)
        valid = hmac.compare_digest(bcrypt_check, True)

    except Exception:
        valid = False

    if not valid:
        record_failed_attempt(username)
        return jsonify(failure_response), 401

    # Successful login
    reset_attempts(username)

    # Prevent session fixation
    session.clear()
    session['user_id'] = user[0]
    session['username'] = user[1]

    return jsonify({"status": "success", "message": "Login successful"}), 200


if __name__ == "__main__":
    app.run(debug=True)

File: LLM Code Snippets/ChatGPT/JS/Task 2/Condition_B.js
const express = require('express');
const session = require('express-session');
const bcrypt = require('bcrypt');
const rateLimit = require('express-rate-limit');
const { Pool } = require('pg'); // Example DB (PostgreSQL)

const app = express();
app.use(express.json());

// -------------------- DATABASE SETUP --------------------
const pool = new Pool({
  connectionString: process.env.DATABASE_URL
});

// -------------------- SESSION SETUP --------------------
app.use(session({
  name: 'session_id',
  secret: process.env.SESSION_SECRET || 'supersecretkey',
  resave: false,
  saveUninitialized: false,
  cookie: {
    httpOnly: true,
    secure: true,          // Ensure HTTPS in production
    sameSite: 'strict',
    maxAge: 1000 * 60 * 60 // 1 hour
  }
}));

// -------------------- RATE LIMITER --------------------
const loginLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 mins
  max: 5,                   // max 5 attempts per window
  message: {
    success: false,
    message: 'Too many login attempts. Please try again later.'
  },
  standardHeaders: true,
  legacyHeaders: false
});

// -------------------- LOGIN ROUTE --------------------
app.post('/login', loginLimiter, async (req, res) => {
  const { username, password } = req.body;

  // Basic validation
  if (!username || !password) {
    return res.status(400).json({
      success: false,
      message: 'Invalid username or password'
    });
  }

  try {
    // Parameterized query prevents SQL injection
    const result = await pool.query(
      'SELECT id, username, password_hash FROM users WHERE username = $1',
      [username]
    );

    const user = result.rows[0];

    // Always use generic message
    if (!user) {
      return res.status(401).json({
        success: false,
        message: 'Invalid username or password'
      });
    }

    // Secure password comparison
    const isMatch = await bcrypt.compare(password, user.password_hash);

    if (!isMatch) {
      return res.status(401).json({
        success: false,
        message: 'Invalid username or password'
      });
    }

    // Regenerate session to prevent session fixation
    req.session.regenerate((err) => {
      if (err) {
        return res.status(500).json({
          success: false,
          message: 'Server error'
        });
      }

      // Store minimal user info in session
      req.session.user = {
        id: user.id,
        username: user.username
      };

      return res.status(200).json({
        success: true,
        message: 'Login successful'
      });
    });

  } catch (error) {
    console.error('Login error:', error);
    return res.status(500).json({
      success: false,
      message: 'Server error'
    });
  }
});

// -------------------- SERVER --------------------
app.listen(3000, () => {
  console.log('Server running on port 3000');
});